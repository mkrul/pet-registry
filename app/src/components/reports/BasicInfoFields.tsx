import React from "react";
import { TextField, SelectChangeEvent } from "@mui/material";
import { IReportForm } from "../../types/Report";

const commonInputStyles = {
  backgroundColor: "white",
  "& .MuiOutlinedInput-root": {
    backgroundColor: "white"
  }
};

interface BasicInfoFieldsProps {
  formData: IReportForm;
  onInputChange?: (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement> | SelectChangeEvent
  ) => void;
  readOnly?: boolean;
}

export const BasicInfoFields: React.FC<BasicInfoFieldsProps> = ({
  formData,
  onInputChange,
  readOnly
}) => {
  return (
    <>
      <TextField
        label="Title"
        name="title"
        value={formData.title || ""}
        onChange={onInputChange}
        variant="outlined"
        fullWidth
        disabled={readOnly}
        sx={commonInputStyles}
      />

      <TextField
        label="Description"
        name="description"
        value={formData.description || ""}
        onChange={onInputChange}
        multiline
        rows={4}
        variant="outlined"
        fullWidth
        disabled={readOnly}
        sx={commonInputStyles}
      />

      <TextField
        label="Pet's Name (leave blank if not known)"
        name="name"
        value={formData.name || ""}
        onChange={onInputChange}
        variant="outlined"
        fullWidth
        disabled={readOnly}
        sx={commonInputStyles}
      />
    </>
  );
};
