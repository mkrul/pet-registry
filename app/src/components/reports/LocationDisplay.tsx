import React from "react";

interface LocationDisplayProps {
  area: string;
  state: string;
  country: string;
}

export const LocationDisplay: React.FC<LocationDisplayProps> = ({ area, state, country }) => {
  return (
    <div className="mt-4">
      <h2 className="text-lg font-medium text-gray-900 mb-2">Location</h2>
      <div className="grid grid-cols-3 gap-4">
        <div>
          <div className="text-md text-gray-500 mb-1">Area</div>
          <div className="p-2 bg-gray-50 border border-gray-200 rounded-md text-gray-700 truncate">
            {area || "No area selected"}
          </div>
        </div>
        <div>
          <div className="text-md text-gray-500 mb-1">State</div>
          <div className="p-2 bg-gray-50 border border-gray-200 rounded-md text-gray-700 truncate">
            {state || "No state selected"}
          </div>
        </div>
        <div>
          <div className="text-md text-gray-500 mb-1">Country</div>
          <div className="p-2 bg-gray-50 border border-gray-200 rounded-md text-gray-700 truncate">
            {country || "No country selected"}
          </div>
        </div>
      </div>
    </div>
  );
};
