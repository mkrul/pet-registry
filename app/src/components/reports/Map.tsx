import React from "react";
import { MapContainer, TileLayer, Marker, useMapEvents } from "react-leaflet";
import "leaflet/dist/leaflet.css";
import L from "leaflet";

interface MapProps {
  onLocationSelect: (location: {
    latitude: number;
    longitude: number;
    city: string;
    state: string;
    country: string;
  }) => void;
}

const LocationMarker: React.FC<{ onLocationSelect: MapProps["onLocationSelect"] }> = ({
  onLocationSelect
}) => {
  const [position, setPosition] = React.useState<L.LatLng | null>(null);

  useMapEvents({
    click(e) {
      setPosition(e.latlng);
      fetchLocationDetails(e.latlng.lat, e.latlng.lng, onLocationSelect);
    },
    dragend() {
      if (position) {
        fetchLocationDetails(position.lat, position.lng, onLocationSelect);
      }
    }
  });

  return position === null ? null : <Marker position={position} draggable={true} />;
};

const fetchLocationDetails = async (
  lat: number,
  lng: number,
  onLocationSelect: MapProps["onLocationSelect"]
) => {
  try {
    const response = await fetch(
      `https://nominatim.openstreetmap.org/reverse?format=json&lat=${lat}&lon=${lng}&zoom=18&addressdetails=1`
    );
    const data = await response.json();
    const address = data.address;
    onLocationSelect({
      latitude: lat,
      longitude: lng,
      city: address.city || address.town || address.village || "",
      state: address.state || "",
      country: address.country || ""
    });
  } catch (error) {
    console.error("Error fetching location details:", error);
  }
};

const Map: React.FC<MapProps> = ({ onLocationSelect }) => {
  return (
    <div className="w-full h-[400px] rounded-lg overflow-hidden border border-gray-300">
      <MapContainer center={[37.0902, -95.7129]} zoom={4} style={{ width: "100%", height: "100%" }}>
        <TileLayer
          url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
          attribution='Â© <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
        />
        <LocationMarker onLocationSelect={onLocationSelect} />
      </MapContainer>
    </div>
  );
};

export default Map;
