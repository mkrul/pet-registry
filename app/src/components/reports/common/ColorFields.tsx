import React from "react";
import { ColorFieldsProps } from "../../../types/Report";
import { AddFieldButton } from "../../common/AddFieldButton";
import { AdditionalFieldSet } from "../../common/AdditionalFieldSet";
import { ColorSearch } from "../../common/ColorSearch";
import { commonInputStyles } from "../../../styles/commonStyles";
import { Alert } from "@mui/material";

export const ColorFields: React.FC<ColorFieldsProps> = ({
  formData,
  showColor2,
  showColor3,
  setShowColor2,
  setShowColor3,
  onColor2Add,
  onColor3Add,
  onColor2Remove,
  onColor3Remove,
  isLoading,
  handleColor1Change,
  handleColor2Change,
  handleColor3Change,
  error
}) => {
  return (
    <div className="space-y-6">
      <div className="space-y-2">
        <label className="text-lg font-medium text-gray-900">Colors:</label>
        <div className="flex-grow">
          <ColorSearch
            value={formData.color1}
            onChange={handleColor1Change}
            disabled={isLoading}
            size="medium"
            excludeColors={[formData.color2, formData.color3].filter(Boolean)}
            sx={commonInputStyles}
            required
            error={!!error}
          />
          {error && (
            <Alert severity="error" sx={{ mb: 2 }}>
              {error}
            </Alert>
          )}
        </div>
      </div>

      {!showColor2 && formData.color1 && (
        <AddFieldButton
          onClick={() => setShowColor2(true)}
          disabled={isLoading}
          label="ADD COLOR"
          testId="add-color-button"
        />
      )}

      {showColor2 && (
        <div className="mt-8">
          <div className="mt-6">
            <AdditionalFieldSet
              label="Second Color:"
              onRemove={onColor2Remove}
              disabled={isLoading}
              testId="remove-color-button"
            >
              <ColorSearch
                value={formData.color2 || ""}
                onChange={handleColor2Change}
                disabled={isLoading}
                size="medium"
                excludeColors={[formData.color1, formData.color3].filter(Boolean)}
                sx={commonInputStyles}
              />
            </AdditionalFieldSet>
          </div>
        </div>
      )}

      {showColor2 && !showColor3 && formData.color2 && (
        <AddFieldButton
          onClick={() => setShowColor3(true)}
          disabled={isLoading}
          label="ADD COLOR"
          testId="add-color-button"
        />
      )}

      {showColor3 && (
        <div className="mt-8">
          <div className="mt-6">
            <AdditionalFieldSet
              label="Third Color:"
              onRemove={onColor3Remove}
              disabled={isLoading}
              testId="remove-color-button"
            >
              <ColorSearch
                value={formData.color3 || ""}
                onChange={handleColor3Change}
                disabled={isLoading}
                size="medium"
                excludeColors={[formData.color1, formData.color2].filter(Boolean)}
                sx={commonInputStyles}
              />
            </AdditionalFieldSet>
          </div>
        </div>
      )}
    </div>
  );
};
