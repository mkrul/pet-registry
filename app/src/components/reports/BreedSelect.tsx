import React from "react";
import { getBreedsBySpecies, Species } from "../../lib/reports/breedLists";

interface IBreedSelect {
  species: Species | "";
  value: string;
  onChange: (value: string) => void;
}

const BreedSelect: React.FC<IBreedSelect> = ({ species, value, onChange }) => {
  const breeds = getBreedsBySpecies(species || null);

  return (
    <div className="w-full">
      <select
        name="breed"
        value={value}
        onChange={e => onChange(e.target.value)}
        className={`w-full min-w-[100px] rounded-md border-gray-300 py-2 pl-3 pr-10 text-sm focus:border-blue-500 focus:outline-none focus:ring-blue-500 whitespace-nowrap ${
          !species ? "bg-gray-100 text-gray-400 cursor-not-allowed" : ""
        }`}
        disabled={!species}
      >
        <option value="">Breed</option>
        {breeds.map(breed => (
          <option key={breed} value={breed}>
            {breed}
          </option>
        ))}
      </select>
    </div>
  );
};

export default BreedSelect;
