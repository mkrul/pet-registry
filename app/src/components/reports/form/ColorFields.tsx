import React from "react";
import { FormControl, Select, MenuItem, Button } from "@mui/material";
import { getColorOptions } from "../../../lib/reports/colorList";
import { Close as CloseIcon } from "@mui/icons-material";
import { ColorFieldsProps } from "../../../types/Report";
import { commonInputStyles } from "../../../styles/commonStyles";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons";

export const ColorFields: React.FC<ColorFieldsProps> = ({
  formData,
  showColor2,
  showColor3,
  onInputChange,
  setShowColor2,
  setShowColor3,
  onColor2Add,
  onColor3Add,
  onColor2Remove,
  onColor3Remove,
  isLoading
}) => {
  const getFilteredColorOptions = (selectedColors: (string | null)[]) => {
    const validSelectedColors = selectedColors.filter(
      (color): color is string => color !== null && color !== ""
    );
    return getColorOptions().filter(color => !validSelectedColors.includes(color));
  };

  const menuProps = {
    PaperProps: {
      style: {
        maxHeight: 200
      }
    }
  };

  const createChangeEvent = (name: string, value: string | null) =>
    ({
      target: { name, value }
    }) as React.ChangeEvent<HTMLInputElement>;

  const handleColorConflict = (value: string) => {
    if (value === formData.color2) {
      onInputChange(createChangeEvent("color2", null));
      setShowColor2(false);
    }
    if (value === formData.color3) {
      onInputChange(createChangeEvent("color3", null));
      setShowColor3(false);
    }
  };

  const handleColorChange = (e: any) => {
    handleColorConflict(e.target.value);
    onInputChange(e);
  };

  const handleColor2Change = (e: any) => {
    const { value } = e.target;
    if (value === formData.color1) return;
    if (value === formData.color3) {
      onInputChange(createChangeEvent("color3", null));
      setShowColor3(false);
    }
    onInputChange(e);
  };

  const handleColor3Change = (e: any) => {
    if (e.target.value === formData.color1 || e.target.value === formData.color2) return;
    onInputChange(e);
  };

  const handleColor2Remove = () => {
    onInputChange({ target: { name: "color2", value: null } });
    onColor2Remove?.();
    setShowColor2(false);
  };

  const handleColor3Remove = () => {
    onInputChange({ target: { name: "color3", value: null } });
    onColor3Remove?.();
    setShowColor3(false);
  };

  const handleShowColor2Change = (show: boolean) => {
    if (show) {
      onColor2Add?.();
    } else {
      handleColor2Remove();
    }
    setShowColor2(show);
  };

  const handleShowColor3Change = (show: boolean) => {
    if (show) {
      onColor3Add?.();
    } else {
      handleColor3Remove();
    }
    setShowColor3(show);
  };

  return (
    <div className="space-y-2">
      <label className="text-lg font-medium text-gray-900 mb-2">Colors:</label>
      <div className="space-y-6">
        <div className="flex-grow">
          <FormControl fullWidth>
            <Select
              labelId="color1-label"
              id="color1"
              name="color1"
              value={formData.color1}
              onChange={handleColorChange}
              sx={commonInputStyles}
              MenuProps={menuProps}
            >
              {getColorOptions().map((color, index) => (
                <MenuItem key={`${color}-${index}`} value={color}>
                  {color}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </div>

        {!showColor2 && formData.color1 && (
          <Button
            onClick={() => handleShowColor2Change(true)}
            disabled={isLoading}
            color="primary"
            variant="text"
            className="mt-2"
            sx={commonInputStyles}
            data-testid="add-color-button"
          >
            <div className="flex items-center">
              <FontAwesomeIcon icon={faPlus} className="mr-2 mb-[3px]" />
              <span>ADD COLOR</span>
            </div>
          </Button>
        )}

        {showColor2 && (
          <div className="space-y-2">
            <label className="text-lg font-medium text-gray-900 mb-2 mt-2">Second Color:</label>
            <div className="flex items-center gap-4">
              <div className="flex-grow">
                <FormControl fullWidth>
                  <Select
                    labelId="color2-label"
                    id="color2"
                    name="color2"
                    value={formData.color2 || ""}
                    onChange={handleColor2Change}
                    sx={commonInputStyles}
                    MenuProps={menuProps}
                  >
                    {getFilteredColorOptions([formData.color1]).map((color, index) => (
                      <MenuItem key={`${color}-${index}`} value={color}>
                        {color}
                      </MenuItem>
                    ))}
                  </Select>
                </FormControl>
              </div>
              <Button
                type="button"
                onClick={handleColor2Remove}
                disabled={isLoading}
                color="error"
                variant="text"
                startIcon={<CloseIcon fontSize="medium" />}
                aria-label="Remove Color 2"
                sx={commonInputStyles}
                data-testid="remove-color-button"
              >
                Remove
              </Button>
            </div>
          </div>
        )}

        {showColor2 && !showColor3 && formData.color2 && (
          <Button
            onClick={() => handleShowColor3Change(true)}
            disabled={isLoading}
            color="primary"
            variant="text"
            className="mt-2"
            sx={commonInputStyles}
            data-testid="add-color-button"
          >
            <div className="flex items-center">
              <FontAwesomeIcon icon={faPlus} className="mr-2 mb-[3px]" />
              <span>ADD COLOR</span>
            </div>
          </Button>
        )}

        {showColor3 && (
          <div className="space-y-2">
            <label className="text-lg font-medium text-gray-900 mb-2 mt-2">Third Color:</label>
            <div className="flex items-center gap-4">
              <div className="flex-grow">
                <FormControl fullWidth>
                  <Select
                    labelId="color3-label"
                    id="color3"
                    name="color3"
                    value={formData.color3 || ""}
                    onChange={handleColor3Change}
                    sx={commonInputStyles}
                    MenuProps={menuProps}
                  >
                    {getFilteredColorOptions([formData.color1, formData.color2]).map(
                      (color, index) => (
                        <MenuItem key={`${color}-${index}`} value={color}>
                          {color}
                        </MenuItem>
                      )
                    )}
                  </Select>
                </FormControl>
              </div>
              <Button
                type="button"
                onClick={handleColor3Remove}
                disabled={isLoading}
                color="error"
                variant="text"
                startIcon={<CloseIcon fontSize="medium" />}
                aria-label="Remove Color 3"
                sx={commonInputStyles}
                data-testid="remove-color-button"
              >
                Remove
              </Button>
            </div>
          </div>
        )}
      </div>
    </div>
  );
};
