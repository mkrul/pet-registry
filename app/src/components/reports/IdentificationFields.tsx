import React from "react";
import {
  FormControl,
  TextField,
  InputLabel,
  Select,
  MenuItem,
  SelectChangeEvent,
  Button
} from "@mui/material";
import { IReportForm } from "../../types/Report";
import { Species } from "../../lib/reports/breedLists";
import { getGenderOptions } from "../../lib/reports/genderLists";
import speciesListJson from "../../../../config/species.json";
import BreedSelect from "./BreedSelect";

interface IdentificationFieldsProps {
  formData: IReportForm;
  showBreed2: boolean;
  onInputChange: (
    e:
      | React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
      | SelectChangeEvent
      | { target: { name: string; value: boolean | null } }
  ) => void;
  onBreedChange: (breed: string) => void;
  onBreed2Change: (breed: string) => void;
  onSpeciesChange: (species: Species) => void;
  onShowBreed2Change: (show: boolean) => void;
  isLoading?: boolean;
}

const commonInputStyles = {
  backgroundColor: "white",
  "& .MuiOutlinedInput-root": {
    backgroundColor: "white"
  },
  "& .MuiSelect-select": {
    backgroundColor: "white"
  }
};

export const IdentificationFields: React.FC<IdentificationFieldsProps> = ({
  formData,
  showBreed2,
  onInputChange,
  onBreedChange,
  onBreed2Change,
  onSpeciesChange,
  onShowBreed2Change,
  isLoading
}) => {
  const genderOptions = getGenderOptions();
  const speciesOptions = speciesListJson.options;

  return (
    <>
      <TextField
        label="Microchip ID (leave blank if not known)"
        name="microchipId"
        value={formData.microchipId || ""}
        onChange={onInputChange}
        variant="outlined"
        fullWidth
        sx={commonInputStyles}
      />

      <FormControl fullWidth>
        <InputLabel id="gender-label">Gender (leave blank if not known)</InputLabel>
        <Select
          labelId="gender-label"
          id="gender"
          name="gender"
          value={formData.gender || ""}
          onChange={onInputChange}
          label="Gender"
          sx={commonInputStyles}
        >
          {genderOptions.map((gender, index) => (
            <MenuItem key={index} value={gender}>
              {gender}
            </MenuItem>
          ))}
        </Select>
      </FormControl>

      <FormControl fullWidth>
        <InputLabel id="species-label">Species</InputLabel>
        <Select
          labelId="species-label"
          id="species"
          name="species"
          value={formData.species}
          onChange={e => onSpeciesChange(e.target.value as Species)}
          label="Species"
          sx={commonInputStyles}
        >
          {speciesOptions.map((species, index) => (
            <MenuItem key={`${species}-${index}`} value={species}>
              {species}
            </MenuItem>
          ))}
        </Select>
      </FormControl>

      <BreedSelect
        species={formData.species as Species | ""}
        value={formData.breed1}
        onChange={onBreedChange}
        variant="form"
        label="Breed 1"
      />

      {!showBreed2 && formData.breed1 && (
        <Button
          onClick={() => onShowBreed2Change(true)}
          disabled={isLoading}
          color="primary"
          variant="text"
          className="mt-2"
        >
          + ADD ANOTHER BREED
        </Button>
      )}

      {showBreed2 && (
        <div className="flex items-center gap-2">
          <BreedSelect
            species={formData.species as Species | ""}
            value={formData.breed2 || ""}
            onChange={onBreed2Change}
            variant="form"
            label="Breed 2"
            excludeBreeds={[formData.breed1]}
          />
          <button
            type="button"
            onClick={() => onShowBreed2Change(false)}
            className="text-red-600 hover:text-red-700 p-1 ml-1"
            disabled={isLoading}
            aria-label="Remove Breed 2"
          >
            âœ•
          </button>
        </div>
      )}
    </>
  );
};
