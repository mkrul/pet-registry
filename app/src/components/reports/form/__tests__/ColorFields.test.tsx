import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/vitest";
import { ColorFields } from "../ColorFields";
import type { ColorFieldsProps } from "../../../../types/Report";
import { getColorOptions } from "../../../../lib/reports/colorList";
import { act } from "react";

const mockFormData = {
  title: "Test Title",
  description: "Test Description",
  name: "Test Pet Name",
  species: "Dog",
  breed1: "Labrador",
  breed2: null,
  color1: "Black" as string,
  color2: "" as string,
  color3: "" as string,
  gender: "Male",
  microchipId: "123456",
  latitude: 0,
  longitude: 0,
  area: "Test Area",
  state: "Test State",
  country: "Test Country",
  image: {
    id: "1",
    url: "/test-image.jpg",
    thumbnailUrl: "/test-image-thumb.jpg",
    variantUrl: "/test-image.jpg",
    filename: "test-image.jpg",
    publicId: "test-image"
  }
};

describe("ColorFields", () => {
  const mockOnInputChange = vi.fn();
  const mockOnShowColor2Change = vi.fn();
  const mockOnShowColor3Change = vi.fn();

  const mockProps: Required<ColorFieldsProps> = {
    formData: mockFormData,
    showColor2: false,
    showColor3: false,
    onInputChange: mockOnInputChange,
    onShowColor2Change: mockOnShowColor2Change,
    onShowColor3Change: mockOnShowColor3Change,
    isLoading: false
  };

  beforeEach(() => {
    mockOnInputChange.mockClear();
    mockOnShowColor2Change.mockClear();
    mockOnShowColor3Change.mockClear();
  });

  describe("Rendering", () => {
    it("should match snapshot", () => {
      const { container } = render(<ColorFields {...mockProps} />);
      expect(container).toMatchSnapshot();
    });

    it("should render color1 select with all color options", () => {
      render(<ColorFields {...mockProps} />);
      const select = screen.getByLabelText(/color 1/i);
      expect(select).toBeDefined();

      const input = within(select.parentElement!).getByRole("combobox");
      expect(input.textContent).toBe("Black");
    });

    it("should not show color2 and color3 fields initially", () => {
      render(<ColorFields {...mockProps} />);
      expect(screen.queryByRole("combobox", { name: /color 2/i })).toBeNull();
      expect(screen.queryByRole("combobox", { name: /color 3/i })).toBeNull();
    });

    it("should show 'Add Another Color' button when color1 is selected", () => {
      render(<ColorFields {...mockProps} />);
      const addButton = screen.getByText(/add another color/i, { selector: "button" });
      expect(addButton).toBeDefined();
    });
  });

  describe("Color Selection Logic", () => {
    it.each(getColorOptions())("should allow selecting %s as color1", async color => {
      render(<ColorFields {...mockProps} />);

      const select = screen.getByLabelText(/color 1/i);
      expect(select).toBeDefined();

      await act(async () => {
        const changeEvent = {
          target: { name: "color1", value: color }
        };
        mockProps.onInputChange(changeEvent as React.ChangeEvent<HTMLInputElement>);
      });

      expect(mockOnInputChange).toHaveBeenCalledWith(
        expect.objectContaining({
          target: expect.objectContaining({
            name: "color1",
            value: color
          })
        })
      );
    });

    it("should prevent duplicate color selection", async () => {
      render(
        <ColorFields
          {...mockProps}
          showColor2={true}
          formData={{ ...mockFormData, color1: "Black", color2: "White" }}
        />
      );

      const select = screen.getByLabelText("Color 2");
      expect(select).toBeDefined();

      const selectElement = within(select.parentElement!).getByRole("combobox");
      await act(async () => {
        await userEvent.click(selectElement);
      });

      const options = screen.queryByRole("option", { name: "Black" });
      expect(options).toBeNull();
    });

    it("should clear color3 when selecting same color in color2", async () => {
      render(
        <ColorFields
          {...mockProps}
          showColor2={true}
          showColor3={true}
          formData={{ ...mockFormData, color2: "White", color3: "Brown" }}
        />
      );

      await act(async () => {
        mockProps.onInputChange({
          target: { name: "color2", value: "Brown" }
        } as React.ChangeEvent<HTMLInputElement>);

        mockProps.onInputChange({
          target: { name: "color3", value: "" }
        } as React.ChangeEvent<HTMLInputElement>);
      });

      const calls = mockOnInputChange.mock.calls;
      expect(calls[0][0].target).toEqual({ name: "color2", value: "Brown" });
      expect(calls[1][0].target).toEqual({ name: "color3", value: "" });
    });
  });

  describe("Add/Remove Color Fields", () => {
    it("should show color2 field when add button is clicked", async () => {
      render(<ColorFields {...mockProps} />);
      const addButton = screen.getByText(/add another color/i, { selector: "button" });

      await act(async () => {
        await userEvent.click(addButton);
      });

      expect(mockOnShowColor2Change).toHaveBeenCalledWith(true);
    });

    it("should remove color2 field when remove button is clicked", async () => {
      render(<ColorFields {...mockProps} showColor2={true} />);
      const removeButton = screen.getByLabelText(/remove color 2/i);

      await act(async () => {
        await userEvent.click(removeButton);
      });

      expect(mockOnShowColor2Change).toHaveBeenCalledWith(false);
    });

    it("should disable add/remove buttons when loading", () => {
      render(<ColorFields {...mockProps} showColor2={true} isLoading={true} />);

      // Use queryByRole to check if buttons exist and are disabled
      const addButton = screen.queryByRole("button", {
        name: /add another color/i,
        hidden: true
      });
      const removeButton = screen.queryByRole("button", {
        name: /remove color 2/i,
        hidden: true
      });

      // Check if buttons are either disabled or not rendered when loading
      if (addButton) {
        expect(addButton).toBeDisabled();
      } else {
        expect(addButton).toBeNull();
      }

      if (removeButton) {
        expect(removeButton).toBeDisabled();
      } else {
        expect(removeButton).toBeNull();
      }
    });
  });

  describe("Error Handling", () => {
    it("should handle empty values", () => {
      render(
        <ColorFields
          {...mockProps}
          formData={{
            ...mockFormData,
            color1: "",
            color2: "",
            color3: ""
          }}
        />
      );

      const select = screen.getByLabelText(/color 1/i);
      const input = within(select.parentElement!).getByRole("combobox");
      expect(input.textContent?.replace(/\u200B/g, "")).toBe("");
    });
  });
});
