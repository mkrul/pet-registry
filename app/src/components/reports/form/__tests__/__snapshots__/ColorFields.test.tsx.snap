// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorFields > Rendering > should match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-ytlejw-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      id="color1-label"
    >
      Color 1
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-fevcd3-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-controls=":r0:"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="color1-label color1"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="color1"
        role="combobox"
        tabindex="0"
      >
        Black
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        name="color1"
        tabindex="-1"
        value="Black"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w1u3ce"
        >
          <span>
            Color 1
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary mt-2 css-1588512-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    + ADD COLOR
  </button>
</div>
`;
